tree.R

```diff
@@ -165,112 +176,211 @@ build_tree <- function(df, filter_prevalence, filter_mean_abundance) {
+      loser_names <- append(loser_names, competitor$name)
+    }
+  }
+  outcome_str <- sprintf("winners: %s; losers: %s", paste(winner_names, collapse = ","), paste(loser_names, collapse = ","))
+
+  # now actually mark the results, including outcome string
   # mark winners/losers of parent and descendants
   # also mark the non-winners as rf losers
-  for (descendant in descendant_winners) {
-    if (descendant$id %in% rf_winners) {
-      descendant$winner <- TRUE
+  for (competitor in competitors) {
+    if (competitor$id %in% rf_winners) {
+      competitor$outcomes <- append(competitor$outcomes, sprintf("win: rf winner, %s", outcome_str))
+      competitor$winner <- TRUE
     } else {
-      descendant$winner <- FALSE
-      descendant$lost_rf <- TRUE
+      competitor$outcomes <- append(competitor$outcomes, sprintf("loss: rf loser, %s", outcome_str))
+      competitor$winner <- FALSE
+      competitor$lost_rf <- TRUE
     }
   }
-  
+
   return()
 }
 
+# compete all winners, updating the tree in the process
+# TODO: combine the overlaps in this code with the code in the function above
+compete_all_winners <- function(tree, metadata, sample_fraction, ncores) {
+  # all vs all competition with winners
+  # skipped rows have winner = FALSE so won't appear in this list
+  competitors <- get_descendant_winners(tree, tree$height)
+  if (length(all_winners) == 0) {
+    return(tree)
+  }
+
+  # all winners into a transposed data frame
+  df <- data.frame()
+  for (winner in competitors) {
+    df <- rbind(df, winner$abundance)
+  }
+  transposed <- as.data.frame(t(df))
+
+  # compete here
+  # return list of winner ids
+  # right now does nothing, all are marked as winners
+  winners <- lapply(competitors, function(node) node$id)
+
+  # TODO: so much duplication below
+
+  # generate winner and loser name lists from the competitors (parent and non-correlated descendants), using the outcome
+  # build ahead of time so that a summary can be provided in outcomes
+  # this can be sped up and done in a single loop if outcomes are not needed
+  winner_names <- list()
+  loser_names <- list()
+  for (competitor in competitors) {
+    if (competitor$id %in% rf_winners) {
+      winner_names <- append(winner_names, competitor$name)
+    } else {
+      loser_names <- append(loser_names, competitor$name)
+    }
+  }
+  outcome_str <- sprintf("winners: %s; losers: %s", paste(winner_names, collapse = ","), paste(loser_names, collapse = ","))
+
+  # now actually mark the results, including outcome string
+  # mark winners/losers of parent and descendants
+  # also mark the non-winners as rf losers
+  for (competitor in competitors) {
+    if (competitor$id %in% rf_winners) {
+      competitor$outcomes <- append(competitor$outcomes, sprintf("win: rf winner, %s", outcome_str))
+      competitor$winner <- TRUE
+    } else {
+      competitor$outcomes <- append(competitor$outcomes, sprintf("loss: rf loser, %s", outcome_str))
+      competitor$winner <- FALSE
```

Comment:
should the all-vs-all loser still be marked as winners? Or maybe there needs to be a new field